Nested tensor forward only chunk operations

drisspg opened the pull request:
# Summary

Taking over this pr: https://github.com/pytorch/pytorch/pull/83736

Adding support for chunk without autograd support

drisspg committed:
Support chunk and add tests

cpuhrsch reviewed (commented):
So what happens why I try to use autograd together with these ops?

cpuhrsch commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -2175,10 +2163,6 @@ Tensor slice(
     auto quantizer = create_subtensor_quantizer(self, false, start_val, end_val, dim, step);
     result = as_strided_qtensorimpl(self, sizes, strides, storage_offset, quantizer);
   } else {
-    // NB: it is extremely important to perform a redispatch here for
```
Why was this removed?

drisspg replied:
whoops I think this comment was added after Joels pr I will add back

drisspg commented:
> So what happens why I try to use autograd together with these ops?

For ops split, split_with_sizes since the registered derivative formula calls `self.size()` to save for backward we get an error:

```
import torch
a = torch.nested_tensor([torch.randn(2, 3, 6, requires_grad=True), torch.randn(4, 5, 6, requires_grad=True)])
b = torch.split(a, 2, dim=-1)

RuntimeError: Internal error: NestedTensorImpl doesn't support sizes. Please file an issue on https://github.com/pytorch/nestedtensor
```

For the composite ops like, chunk, narrow even though they are composite the same error arises because they all end up hitting split or split with sizes that has the `.sizes` call.

drisspg commented:
> > So what happens why I try to use autograd together with these ops?
> 
> For ops split, split_with_sizes since the registered derivative formula calls `self.size()` to save for backward we get an error:
> 
> ```
> import torch
> a = torch.nested_tensor([torch.randn(2, 3, 6, requires_grad=True), torch.randn(4, 5, 6, requires_grad=True)])
> b = torch.split(a, 2, dim=-1)
> 
> RuntimeError: Internal error: NestedTensorImpl doesn't support sizes. Please file an issue on https://github.com/pytorch/nestedtensor
> ```
> 
> For the composite ops like, chunk, narrow even though they are composite the same error arises because they all end up hitting split or split with sizes that has the `.sizes` call.

@cpuhrsch 
Follow up pr would be to use the Mikayla's derivative.yaml notation to register the derivative with slice and split/split_with_sizes.

cpuhrsch commented:
@drisspg - let's register "NotImplemented" as the derivative in a follow up. Let's at least have a good error message.

cpuhrsch commented on test/test_nestedtensor.py:
```diff
@@ -549,6 +552,109 @@ def test_nested_tensor_indexing(self, device, dtype):
+
+        # Error case: slicing jagged dimension
+        with self.assertRaisesRegex(RuntimeError, 'dimension 1 is irregular'):
+            x.narrow(1, 1, 2)
```
Just for future reference, this doesn't *have* to fail. We could easily narrow along an irregular dimension as long as there's enough data, it'll just become regular as a result.

cpuhrsch reviewed (approved):
Neat!

drisspg commented:
@pytorchbot merge -l

mikaylagawarecki commented on test/test_nestedtensor.py:
```diff
@@ -549,6 +552,109 @@ def test_nested_tensor_indexing(self, device, dtype):
+        b_nt = [a_chunks[1], b_chunks[1], c_chunks[1]]
+        c_nt = [a_chunks[2], b_chunks[2], c_chunks[2]]
+
+        nt = torch.nested_tensor([a, b, c])
```
@drisspg This might landrance with the nested_tensor/as_nested_tensor PR )":

Would you be open to waiting for that to land before landing this? Asking because I'm landing that via phabricator due to internal changes that will break tests and phabricator tests take extremely long to run (and are almost done for the latest rebase right now!)

Alternatively, could we land this test case specifically in a separate PR (I could open a PR that adds this test back after my PR lands)? If not I'll wait for this to land and then rebase on my end (but just a little nervous about getting everything in hahah)

drisspg replied:
Yeah of course, I just canceled my merge job. Just Ping me and let mw know when I can land. I assume it will be tomorrow after the diff train with yours lands?

mikaylagawarecki replied:
hm mine is not in diff train bcos i'm clicking land on phabricator! If nothing goes wrong it should be landed before tmr morning, will ping you!

drisspg commented:
@pytorchbot merge

drisspg commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -658,11 +658,11 @@ Tensor clone_nested(
     const Tensor& self_buffer = self_ptr->get_buffer(),
         sizemat = self_ptr->get_nested_size_tensor();
-    Tensor output_buffer = at::empty_like(self_buffer);
+    Tensor output_buffer = at::empty(self.numel(), self_buffer.options());
```
I included the updates to this clone function here

drisspg commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1280,6 +1280,9 @@
   variants: function, method
   device_check: NoCheck
   device_guard: False
+  dispatch:
```
@cpuhrsch  when building from source with debug=0:

```
TORCH_INTERNAL_ASSERT(
          aliased_input.storage().is_alias_of(aliased_output.storage()),
            op_name);
````
The above line will be added to the codgened kernel for nested tensor because of this signature. I added a comment on the test. I really don't like this because it makes dev experience worse but I couldn't figure out a nice way around it.

drisspg commented:
@cpuhrsch removed all by the restricted chunk and the small update to clone

drisspg committed:
lint baby

drisspg committed:
rebase changes

drisspg committed:
update tests

drisspg commented:
@pytorchbot merge -l

drisspg commented:
@pytorchbot merge
